### ISSUE TITLE
[docs] Base Laravel queue events: JobProcessing, JobProcessed, JobFailed...

### ISSUE THREAD

**Issue by andrewmclagan** [laravel/horizon#348]
It seems that Horizon overrides the Laravel queue events: `JobProcessing, JobProcessed, JobFailed...` is this correct? If so the following would of course not work as per the documentation in the core Laravel docs. 

It should be documented to save devs (like myself) wading through code / debugging test to discover this. I'd be happy to write a docs PR if this is tenable.

[docs link](https://laravel.com/docs/5.6/queues#job-events)

```php
        Queue::before(function (JobProcessing $event) {
            // $event->connectionName
            // $event->job
            // $event->job->payload()
        });
```

These events seem to be replace by [`Laravel\Horizon\Events`](https://github.com/laravel/horizon/tree/1.0/src/Events) is that correct?

**Comment by mfn** [comment]
I'm using this approach since before I introduced Horizon into my stack and it worked before/after without problems.

What is the error/behavior you're seeing that makes you think otherwise?

**Comment by andrewmclagan** [comment]
In our integration test suite for a package that stores metrics in Redis (for distributed performance measuring) everything works fine

class here https://github.com/ethical-jobs/quantify/blob/v1.2.2/src/Reporters/Queues.php#L67

test case here https://github.com/ethical-jobs/quantify/blob/v1.2.2/tests/Integration/Reporters/QueuesTest.php#L39

Everything works fine, tests pass green (try it out). Although when we use the package the events just are not being fired anywhere in the app.

**Comment by mfn** [comment]
One difference I see: I'm using the `Queue`-facade, not `Event`.

I'm using a "global" listener concept here, which is initialized as part of the `AppServiceProvider`:
```php
    public function boot(): void
    {
        $this->initializeQueueEvents();
    }

    protected function initializeQueueEvents()
    {
        Queue::before(function (JobProcessing $event) {
            $job = $event->job;
            $logger = $this->getLoggerForQueue($job->getQueue());
            $logger->info(\get_class($event), $this->getJobDetails($job));
        });

        Queue::after(function (JobProcessed $event) {
            $job = $event->job;
            $logger = $this->getLoggerForQueue($job->getQueue());
            $logger->info(\get_class($event), $this->getJobDetails($job));
        });

        Queue::exceptionOccurred(function (JobExceptionOccurred $event) {
            $job = $event->job;
            $logger = $this->getLoggerForQueue($job->getQueue());
            $details = $this->getJobDetails($job);
            $details['exception'] = $event->exception;
            $logger->error(\get_class($event), $details);
        });

        Queue::failing(function (JobFailed $event) {
            $job = $event->job;
            $logger = $this->getLoggerForQueue($job->getQueue());
            $details = $this->getJobDetails($job);
            $details['exception'] = $event->exception;
            $logger->error(\get_class($event), $details);
        });
    }

```

**Comment by andrewmclagan** [comment]
We were using the Queue Facades also although we swapped it out for the Event Facades thinking it may address the issue. The main issue here that I cant resolve is that the tests pass, they are pretty clear in the scope they are testing. Although once wired up in an actual app the events dont seem to fire. 

The only difference is the app we are testing with has Horizon installed?

**Comment by andrewmclagan** [comment]
Okay I will close this issue as looking deeper Horizon does not screw with normal global queue events. 

The issue we experienced is resolved. It was due to the fact out job tracking logic was tested under a "sync" queue environment not a distributed Redis environment where queue jobs start and complete asynchronously.  

We now setup the event listeners inside the ServiceProvider where they will pick up jobs that a started in a new PHP process (queue worker).

[See the `isJobTracked()` function](https://github.com/ethical-jobs/quantify/blob/v1.2.3/src/Reporters/Queues.php#L72)

[and also the service provider call `listenToQueue()`](https://github.com/ethical-jobs/quantify/blob/v1.2.3/src/ServiceProvider.php#L78)

For those who experience similar issues.
